extensions/verifier/src/verifier/transcript.rs

```diff
@@ -0,0 +1,119 @@
+use std::{
+    io::{self, Read},
+    marker::PhantomData,
+};
+
+use halo2curves_axiom::{
+    bn256::{Fq, Fr as Halo2Fr, G1Affine},
+    Coordinates, CurveAffine,
+};
+use itertools::Itertools;
+use openvm_ecc_guest::algebra::IntMod;
+use openvm_keccak256_guest::keccak256;
+use openvm_pairing_guest::{
+    affine_point::AffineCoords,
+    bn254::{Bn254G1Affine as EcPoint, Fp, Scalar as Fr},
+};
+use snark_verifier::{
+    util::transcript::{Transcript, TranscriptRead},
+    Error,
+};
+
+use super::{
+    loader::{OpenVmLoader, LOADER},
+    traits::{OpenVmEcPoint, OpenVmScalar},
+};
+
+#[derive(Debug)]
+pub struct OpenVmTranscript<C: CurveAffine, S, B> {
+    stream: S,
+    buf: B,
+    _marker: PhantomData<C>,
+}
+
+impl<S> Transcript<G1Affine, OpenVmLoader> for OpenVmTranscript<G1Affine, S, Vec<u8>> {
+    fn loader(&self) -> &OpenVmLoader {
+        &LOADER
+    }
+
+    fn squeeze_challenge(
+        &mut self,
+    ) -> <super::loader::OpenVmLoader as snark_verifier::loader::ScalarLoader<Halo2Fr>>::LoadedScalar
+    {
+        let data = self
+            .buf
+            .iter()
+            .cloned()
+            .chain(if self.buf.len() == 0x20 {
+                Some(1)
+            } else {
+                None
+            })
+            .collect_vec();
+        let hash = keccak256(&data);
+        self.buf = hash.to_vec();
+        OpenVmScalar(Fr::from_be_bytes(&hash), PhantomData)
```

Comment:
currently the openvm executor can accept any bytes32 (and do computation correctly).  But for prover, the Fr::from_be_bytes input must be within Fr range.   

Need `let value = value % modulus::<F>();`